### ISSUE TITLE
Support OpenCL in core and examples

### ISSUE THREAD

**Issue by thawkins** [burn-rs/burn#278]
### Feature description

Provide support for OpenCL as first class capability. 

### Feature motivation

CUDA only supported on NVidia Hardware, other GPU makers are mostly behind OpenCL

**Comment by nathanielsimard** [comment]
Could be linked to #243

**Comment by thawkins** [comment]
> Could be linked to #243
WebGPU is more about portable graphics, its more in the OpenGL space, OpenCL is about accessing the GPU or other accelerators for compute resource, im not sure that these are hitting the same requirements. 

https://www.techlusive.in/apps/google-chrome-gets-webgpu-tech-that-allows-high-performance-3d-graphics-1369760/#:~:text=%E2%80%9CWebGPU%20is%20a%20new%20web,said%20in%20a%20blog%20post.

**Comment by nathanielsimard** [comment]
This is true, however compute shaders can be used for general cross platform GPGPU.

**Comment by antimora** [comment]
@thawkins have you come across a good Rust library that implements kernels for OpenCL? We don't have the specialty for writing kernels. Are there good implementations out there?

The closest that I found is [TVM's C++ implementation that has Rust bindings ](https://github.com/apache/tvm/blob/main/rust/tvm-sys/src/device.rs#L70).

**Comment by thawkins** [comment]
No, im afraid I dont, Im just myself starting to get familiar with CUDA,
OpenCL and ML applications. Im moving from c/c++ to rust, and was looking
for a good framework to learn. All my machines are intel 11th/12th gen with
Xe, hence the interest.

I will however do some research and report back if i find anything.

**Comment by antimora** [comment]
OpenCL is deprecated in macOS. Metal framework is suggested instead. https://developer.apple.com/library/archive/documentation/Performance/Conceptual/OpenCL_MacProgGuide/Introduction/Introduction.html

I don't see much support out there and there is non that I could find for Rust. 

@nathanielsimard @thawkins I would recommend we close this issue and we address the original request of non-nvidia GPU support by doing work in https://github.com/burn-rs/burn/issues/243.

**Comment by thawkins** [comment]
Im ok, with it, as chrome 113 supports webgpu, for the record the opencl3 crate supports opencl. Also rusticl is an opencl impementation of opencl written in rust which is part of the mesa package.

**Comment by nathanielsimard** [comment]
wgpu also targets vulkan and metal, making it a decent choice to build a cross platform backend. 

However, I'm still unsure if it's the best tech to build the most performant GPU backend, we might explore how we can call non-wgpu based kernel implementations from it. 

For example, we might implement most things with wgpu, but call CuBlas for some operations that can leverage tensor cores (without having to translate the memory format). It would be nice to support specialized kernels with compilation flag to better leverage the hardware, while still being cross-platform.

I'm going to close the issue, since I don't think OpenCL will have first class support, but feel free to continue the discussion in the wgpu issue #243. If some other tech seems more promising, we might open another issue then.
